### PR TITLE
Fixes #27380 - Handle non-unique role on creation

### PR DESCRIPTION
**Pull request by xprazak2**
Not verified this really fixes the issue as we do not have a reproducer.

### CODE REVIEW

**Review by ares — COMMENTED** [review]
Tests failures unrelated, I tested from 2 consoles by putting some extra sleep and it worked fine. The only question I have is, should we introduce locking that could potentially lead to deadlocks? Or should we ignore not unique errors, like we do [here](https://github.com/theforeman/foreman/blob/develop/app/services/fact_importer.rb) already?

If I'm the only one, I'm fine with merging this, it's easy to change if it turns to be problematic.

File: app/services/foreman/plugin/role_lock.rb [hunk 1]
```diff
@@ -19,6 +19,7 @@ def register_role(name, permissions, role_registry, description = '')
       def process_role(name, permissions, description = '')
         role = Role.find_by :name => name
         if role
+          role.lock!
```
Comment (ekohl) [inline app/services/foreman/plugin/role_lock.rb]: Can you help me understand when this lock would be released?

Comment (xprazak2) [reply app/services/foreman/plugin/role_lock.rb]: It would be released at the end of the surrounding transaction.

**Comment by lzap** [comment]
Honestly, I don't like it. We need to find out which processes were running in parallel (from puppet logs if we are able to do it) and lock on the process level (e.g. apipie + seed + migration = all one lockfile). Alternatively, this should be fixed in the puppet itself - it should not run these two tasks at the same time at all.

The reason why I don't think this is a good solution is that this line caused the issue for the particular case, we fix it, but there will be another line probably causing similar issue after that. It's probably better to fix it properly.

I don't think this should be a blocker issue - rerunning installer should fix this.

**Comment by ares** [comment]
@lzap this can happen regardless of the installer. The race condition happens when rails env starts. Thus can be two passenger workers starting (two users accessing Foreman) or regular cron rake taks and a passenger start, or installer and passenger start. I don't think process locking would be better.

**Comment by ares** [comment]
To add to the previous reply, running an installer for the first time and seeing it error out is not a great experience, we shouldn't default to tell people to try it again.

I haven't heard an argument against row locking itself, I'll give it one more day and if nothing changes, I'll go ahead and merge.

**Comment by ekohl** [comment]
> To add to the previous reply, running an installer for the first time and seeing it error out is not a great experience, we shouldn't default to tell people to try it again.

Indeed. Users need to trust the installer and seeing it fail breaks the trust in the project. If I evaluate a product and it crashes in the installer, there's a good chance I'll not even try the product.

On a technical level I would also like to implement a test that reruns the installer and expect no changes. For that it needs a solid reliable base.

> I haven't heard an argument against row locking itself, I'll give it one more day and if nothing changes, I'll go ahead and merge.

In practice I've seen some issues with row locking and deadlocks under load. It helps to keep the transaction as small as possible. So not a transaction for a full controller but rather just the code that modifies a part with a lock. I can't tell if that's the case here, but just wanted to say that. Not an objection in general.

**Review by ares — CHANGES_REQUESTED** [review]
> In practice I've seen some issues with row locking and deadlocks under load. It helps to keep the transaction as small as possible. So not a transaction for a full controller but rather just the code that modifies a part with a lock. I can't tell if that's the case here, but just wanted to say that. Not an objection in general.

As far as I know, this method is not triggered from the controller but only from plugin DSL and the lock would happen only in case the role already exist. The transaction finishes with next save (each active record is wrapped in transaction). Now looking at the the method more carefully, that would unlock the role too soon in case description is updated and origin could fail with the same.

When I went through the backtraces in the original report, they leads to line 43 though. The problem is in role creation. I'm afraid the locking here perhaps helps in some case, but not in where two processes tries to create the same role at the same time in the empty DB.

@xprazak2 I think the best thing we can do here is wrapping the `Role.create!` in begin/rescue and ignoring the error if it's uniqueness validation that prevented the save.

```
      def create_plugin_role(name, permissions, description = '')
        Role.ignore_locking do
begin
          role = Role.create! :name => name, :origin => @plugin_id, :description => description
          role.add_permissions!(permissions)
rescue ActiveRecord::RecordNotUnique
  role = Role.find_by_name(name)
end
          role
        end
      end
```

that should fix the issue without

**Comment by xprazak2** [comment]
Updated with changes as suggested.

File: app/services/foreman/plugin/role_lock.rb [hunk 2]
```diff
@@ -40,8 +40,12 @@ def process_role(name, permissions, description = '')
 
       def create_plugin_role(name, permissions, description = '')
         Role.ignore_locking do
-          role = Role.create! :name => name, :origin => @plugin_id, :description => description
-          role.add_permissions!(permissions)
+          begin
+            role = Role.create! :name => name, :origin => @plugin_id, :description => description
```
Comment (ekohl) [inline app/services/foreman/plugin/role_lock.rb:44]: Isn't there something like a get_or_create that also does this?

Comment (ares) [reply app/services/foreman/plugin/role_lock.rb:44]: we do still want to fail hard if the creation fails for other unexpected reasons

Comment (ekohl) [reply app/services/foreman/plugin/role_lock.rb:44]: Fair enough. I do wonder if you want to verify the permissions are correct if the role already exists. Can't really think it through so leaving it as an exercise :)

**Comment by ekohl** [comment]
Looks like the PR title/commit doesn't really reflect the content anymore.

File: app/services/foreman/plugin/role_lock.rb (hunk 2 above)
Comment (ares) [reply app/services/foreman/plugin/role_lock.rb:44]: That could still be in progress of updating, so we may get false positives here.

**Comment by xprazak2** [comment]
I updated commit message and title.

**Review by ares — APPROVED** [review]
Thanks @xprazak2, merging!

**Pull request merged**